pandas/tools/util.py

```diff
@@ -31,14 +31,20 @@ def cartesian_product(X):
     array([1, 2, 1, 2, 1, 2])]
 
     """
```

Comment:
The code indicates it should be at least a list-like (or rather a list-like of list-likes). And `pandas.compat.product` (or `itertools.product`) is somewhat similar. 
I'll fix the docstring and add tests for it. Unless you think we need something stricter.